### PR TITLE
Switch to pytest for testing

### PR DESCRIPTION
**Pull request by bmerry**
This is not a complete conversion - NGC-545 captures more steps to be done - but this eliminates the dependency on nosetests, which is unmaintained and no longer works on Python 3.10.

It's a big change but almost all of it is mechanical changes to assertions. The more interesting stuff is split into separate commits.

### CODE REVIEW

File: src/katsdpcontroller/test/test_master_controller.py [hunk 1]
```diff
@@ -821,26 +827,26 @@ async def test_capture_status_all(self) -> None:
...
         await self.client.request('capture-init', 'product2')
         reply, informs = await self.client.request('capture-status')
-        self.assertEqual(reply, [b'2'])
+        assert reply == [b'2']
         # Need to compare just arguments, because the message objects have message IDs
         inform_args = [tuple(msg.arguments) for msg in informs]
-        self.assertEqual([
+        assert [
             (b'product1', b'idle'),
             (b'product2', b'capturing')
```
Comment (richarms) [inline src/katsdpcontroller/test/test_master_controller.py:835]: `            (b'product2', b'capturing')]`

File: src/katsdpcontroller/test/test_master_controller.py [hunk 2]
```diff
@@ -854,23 +860,23 @@ async def test_product_list_all(self, time_mock) -> None:
...
-        self.assertEqual(reply, [b'2'])
+        assert reply == [b'2']
         # Need to compare just arguments, because the message objects have message IDs
         inform_args = [tuple(msg.arguments) for msg in informs]
-        self.assertEqual([
+        assert [
             (b'product1', b'idle, started at 2005-07-25T23:34:15Z'),
             (b'product2', b'idle, started at 2009-02-13T23:31:30Z')
-        ], inform_args)
+        ] == inform_args
```
Comment (richarms) [inline src/katsdpcontroller/test/test_master_controller.py:869]: `-`

File: src/katsdpcontroller/test/test_master_controller.py [hunk 3]
```diff
@@ -854,23 +860,23 @@ async def test_product_list_all(self, time_mock) -> None:
...
         await self.client.request('product-configure', 'product2', CONFIG)
         reply, informs = await self.client.request('product-list')
-        self.assertEqual(reply, [b'2'])
+        assert reply == [b'2']
         # Need to compare just arguments, because the message objects have message IDs
         inform_args = [tuple(msg.arguments) for msg in informs]
-        self.assertEqual([
+        assert [
             (b'product1', b'idle, started at 2005-07-25T23:34:15Z'),
             (b'product2', b'idle, started at 2009-02-13T23:31:30Z')
```
Comment (richarms) [inline src/katsdpcontroller/test/test_master_controller.py:868]: `+]`

File: src/katsdpcontroller/test/test_master_controller.py [hunk 4]
```diff
@@ -854,23 +860,23 @@ async def test_product_list_all(self, time_mock) -> None:
...
     async def test_product_list_one(self, time_mock) -> None:
         """Test product-list with a subarray_product_id argument"""
         await self.client.request('product-configure', 'product', CONFIG)
         reply, informs = await self.client.request('product-list', 'product')
-        self.assertEqual(reply, [b'1'])
+        assert reply == [b'1']
         # Need to compare just arguments, because the message objects have message IDs
         inform_args = [tuple(msg.arguments) for msg in informs]
-        self.assertEqual([(b'product', b'idle, started at 2005-07-25T23:34:15Z')], inform_args)
+        assert [(b'product', b'idle, started at 2005-07-25T23:34:15Z')] == inform_args
```
Comment (richarms) [inline src/katsdpcontroller/test/test_master_controller.py:879]: `assert inform_args == [(b'product', b'idle, started at 2005-07-25T23:34:15Z')] `

File: src/katsdpcontroller/test/test_product_config.py [hunk 5]
```diff
@@ -52,106 +49,106 @@ class TestRecursiveDiff:
...
+        assert out == 'x.b changed from 2 to 3'
 
 
 class TestOverride:
     """Test :meth:`~katsdpcontroller.product_config.override`."""
 
     def test_add(self) -> None:
         out = product_config.override({"a": 1}, {"b": 2})
-        assert_equal({"a": 1, "b": 2}, out)
+        assert {"a": 1, "b": 2} == out
```
Comment (richarms) [inline src/katsdpcontroller/test/test_product_config.py:80]: On second thought, these read logically either way around.

**Review by richarms — APPROVED** [review]
Looks good Bruce. I tried to check everything, but may have missed a few argument-orders-in-assignment 's.

**Comment by bmerry** [comment]
Thanks Richard, that was a fast review with some good catches in a very long and tedious PR. I've made those changes.

**Pull request merged**